File: app/src/main/java/jp/juggler/subwaytooter/dialog/ActionsDialog.java
package jp.juggler.subwaytooter.dialog;

import android.content.Context;
import android.content.DialogInterface;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AlertDialog;
import android.text.TextUtils;

import java.util.ArrayList;

import jp.juggler.subwaytooter.R;

public class ActionsDialog {
	
	private static class Action {
		@NonNull final CharSequence caption;
		@NonNull final Runnable r;
		
		Action( @NonNull CharSequence caption, @NonNull Runnable r ){
			this.caption = caption;
			this.r = r;
		}
	}
	
	private final ArrayList< Action > action_list = new ArrayList<>();
	
	public ActionsDialog addAction( @NonNull CharSequence caption, @NonNull Runnable r ){
		
		action_list.add( new Action( caption, r ) );
		
		return this;
	}
	
	public ActionsDialog show( @NonNull Context context, @Nullable CharSequence title ){
		CharSequence[] caption_list = new CharSequence[ action_list.size() ];
		for( int i = 0, ie = caption_list.length ; i < ie ; ++ i ){
			caption_list[ i ] = action_list.get( i ).caption;
		}
		AlertDialog.Builder b = new AlertDialog.Builder( context )
			.setNegativeButton( R.string.cancel, null )
			.setItems( caption_list, new DialogInterface.OnClickListener() {
				@Override public void onClick( DialogInterface dialog, int which ){
					if( which >= 0 && which < action_list.size() ){
						action_list.get( which ).r.run();
					}
				}
			} );
		
		if( ! TextUtils.isEmpty( title ) ) b.setTitle( title );
		
		b.show();
		
		return this;
	}
}

File: app/src/main/java/jp/juggler/subwaytooter/dialog/ActionsDialog.kt
package jp.juggler.subwaytooter.dialog

import android.content.Context
import jp.juggler.util.coroutine.cancellationException
import jp.juggler.util.data.notEmpty
import jp.juggler.util.ui.dismissSafe
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.suspendCancellableCoroutine
import java.util.*
import kotlin.coroutines.resumeWithException

@Suppress("MatchingDeclarationName")
class ActionsDialogInitializer(
    val title: CharSequence? = null,
) {
    class Action(val caption: CharSequence, val action: suspend () -> Unit)

    val list = ArrayList<Action>()

    fun action(caption: CharSequence, action: suspend () -> Unit) {
        list.add(Action(caption, action))
    }

    @OptIn(ExperimentalCoroutinesApi::class)
    suspend fun showSuspend(context: Context): Action =
        suspendCancellableCoroutine { cont ->
            val dialog = android.app.AlertDialog.Builder(context).apply {
                title?.notEmpty()?.let { setTitle(it) }
                setNegativeButton(android.R.string.cancel, null)
                setItems(list.map { it.caption }.toTypedArray()) { d, i ->
                    if (cont.isActive) cont.resume(list[i]) {_,_,_->}
                    d.dismissSafe()
                }
                setOnDismissListener {
                    if (cont.isActive) cont.resumeWithException(cancellationException())
                }
            }.create()
            cont.invokeOnCancellation { dialog.dismissSafe() }
            dialog.show()
        }
}

suspend fun Context.actionsDialog(
    title: String? = null,
    init: suspend ActionsDialogInitializer.() -> Unit,
) {
    ActionsDialogInitializer(title)
        .apply { init() }
        .showSuspend(this)
        .action.invoke()
}
